### ISSUE TITLE
Deadlocks with semaphores and module tasks

### ISSUE THREAD

**Issue by ljwo** [taskflow/taskflow#647]
**Taskflow deadlocks in certain scenarios involving semaphores**
When a `tf::Semaphore` is acquired or a module task started, the following task cannot start until the same thread on which `.acquire` was called or the module task was started returns, as this thread calls `_corun_until`. If that particular thread later happens to steal a task that requires a semaphore that is already locked and released only after the task with the `.acquire` call or the module task, it will never return, which results in a deadlock.

**To Reproduce**
Two examples that mostly deadlock are here: https://github.com/ljwo/taskflow-semaphores 
The one with two semaphores and no module task has this simple structure:
![s2](https://github.com/user-attachments/assets/73587c50-9e38-4281-b507-63d1696845fe)
where the two workers usually start with `f1` and `f2` and `sema1` is locked for long by `f1`, while the thread that started `f2` gets to steal since it cannot acquire `sema1`, so it tries to run `f3`, but `f3` wants `sema2` which had been acquired up the call stack. Of course, if the lib started to run `f2` and `f3`, there would be no problem, and it sometimes does. 

The other example is a bit more contrived:
![s6](https://github.com/user-attachments/assets/d757b5a2-888b-4bd8-843a-c36f33c0c1d6)
as we need to have a thread go stealing from inside a module task. So in this example there are three workers, two top-level tasks that are started, and the delays on works are set up so that the thread that started the module task goes stealing because `mod_f1` is not ready when `mod_f2` has finished, while in the other branch, hopefully(?), the task waiting up for grabs is `f5_acquire_sema`. If things go as described, deadlock ensues.

**Comment by tsung-wei-huang** [comment]
@ljwo  thank you for pointing this out. Indeed, this is not a bug but relatively how Taskflow defines its semaphore usage. In your case, there is no general solution that can avoid the deadlock since the deadlock always has a chance to exist in your graph when the situation happens "it tries to run `f3`, but `f3` wants `sema2` which had been acquired up the call stack."

In such a situation, the worker cannot release `sema2` from its call stack because this will violate how your application task graph defines the acquisition and release of `sema2`. So, the current implementation requires users to ensure their application task graphs are deadlock-free.

Does this make sense to you?

**Comment by ljwo** [comment]
I am afraid to say that I am not happy with your answer. I obviously know of the dangers of getting myself into deadlocks when using semaphores and I carefully avoided them when specifying the execution graphs. 

In the first example, one line of execution (f1) just acquires and releases `sema1`. The third line of execution (f3) just acquires and releases `sema2`. Both lines have no preconditions and should be guaranteed to finish. 

The second line of execution acquires `sema2` and then `sema1`, this one is more complicated, but this is the only line of execution in the entire graph that needs two semaphores, so the graph has neither AB BA nor ABA specified by the user anywhere. 

When f2 enters execution and cannot acquire `sema1` then that line should be just put on hold until f1 finishes (which should be guaranteed) and then proceed. Instead, taskflow **of its own will** chose to turn f2 into an ABA scenario. And even picking the second A as stolen task could be benign if taskflow figured out it had an unsatisfied precondition (free semaphore) instead of actually running it, put it back into the queue and started looking for some other task. 

Same goes for the second scenario. There is just one semaphore in the graph, with two parallel execution lines that need it, so only one can execute at a time. The lines do not explicitly transfer execution one into the other. This is the example on page 1 of any semaphore manual. If you believe it's a deadlock of my own brewing, please explain.

I am not a newcomer, I have been working with taskflow for the last three years and my cv goes further back than I care to admit. Those two bugs that I opened are field experience reports from a real shop that tries to use taskflow for process control, where the machine just froze on us, hopefully in tests and not in production, which cost weeks of log analysis, distilled into cases that fail usually and not once per month. 

My understanding is that I can do precious **nothing** on the user side to prevent it, short of getting rid of module tasks altogether and making sure that the entire graph uses at most one semaphore, which is a bit drastic.

We considered to contribute to taskflow to solve those issues on the lib side, instead of taking steps to use something else, but we lack important insights. The semaphore implementation in previous releases was different and, as far as I understand, did not start the task if the precondition was not satisfied, so was immune to the problems described here. We do not understand the reasons this was changed - is it documented anywhere? 

As for the module tasks, they have been broken all along. As far as we understand, to solve the issue one would have to perform the join on the subgraph not by going up the call stack of a particular worker thread but by having an injected task that would happen after all the tasks of the subgraph, and so, could be executed by any worker. This is what we can do on the user side to get rid of module tasks anyway. But we do not fully understand the interactions with the features of taskflow we have not been using on a daily basis, eg. dynamic tasking, gpu etc.

**Comment by tsung-wei-huang** [comment]
@ljwo - thank you for the response and I do appreciate your use of Taskflow. Please don't get me wrong. I was trying to help and hopefully we can come up with something better.

Let's focus on the first example first and see if we can agree on the fact that the graph is not deadlock-free. What I was emphasizing is that your graph always has a chance to incur deadlock - when the worker who runs `f2_acquire_sema2` and `f2_acquire_sema1_under_sema2` switches to run `f3_acquire_and_release_sema2`. This basically is the same as a thread recursively locking a `std::mutex` instead of `std::recursive_mutex`. Even with one worker, your graph still has a chance of deadlock depending on the execution order of tasks by the worker (e.g., `f2_acquire_sema2` -> `f2_acquire_sema1_under_sema2` -> `f3_acquire_and_release_sema2`).

Does this make sense to you? - your first graph always has a chance to incur deadlock. 

BTW, I assume you create a `tf::Semaphore` with the count initialized to `1`.

**Comment by ljwo** [comment]
Sorry for the delay in my replying, I was busy putting out the fire this has sparked in the project. Module tasks are gone now.

What you write is a nice rephrasing of why this deadlocks, but I still refuse to admit that this is a user side problem. (Yes, all semaphores considered are initially set up to `1`)

As far as I understand what taskflow claims to achieve, there should be no dependency on particular threads in the way the tasks in the dependency graph are executed, other than there may not be enough free execution resources at a particular moment to execute tasks that are otherwise ready to execute. 

So let's see what's ready to execute and what are the resources if things go the way I cajoled them to (taskflow's behaviour is indeterministic, but I'll ignore the other scenarios).

### Let's start by example 1.

There are two workers. 

They take up `f2_acquire_sema2` and `f1_acquire_and_release_sema1` and are both busy, so `f3_acquire_and_release_sema2` has to wait despite being ready to execute.

When `f2_acquire_sema2` finishes to execute, there are two possible interpretations of what is ready to execute. For sure `f3_acquire_and_release_sema2` is ready, but whether `f2_acquire_sema1_under_sema2` is ready is debatable. Its precondition (free `sema1`) is not satisfied. But the one free worker takes it up anyway, and does not make it back ready to run after figuring out it cannot be executed. Instead, it says "I, worker `k` in my very person, will do that when I return to this place in my call stack", and goes out to execute `f3_acquire_and_release_sema2`. This leaves `f2_acquire_sema1_under_sema2` in a limbo state. It is neither ready, nor under execution. That state cannot be described by the dependencies on the graph.

Next, when the other worker is done with `f1_acquire_and_release_sema1`, it could in principle execute `f2_acquire_sema1_under_sema2`, this task has all the strong dependencies ticked, and could be attempted, and maybe even executed by the free execution resource if the semaphore can be acquired (it can, by now). But that task is in the limbo state and is not available for execution.

As far as I can see, this used to work differently in previous releases of taskflow. This code:
```
   if(node->_semaphores && !node->_semaphores->to_acquire.empty()) {
    SmallVector<Node*> nodes;
    if(!node->_acquire_all(nodes)) {
      _schedule(worker, nodes);
      return;
    }
    node->_state.fetch_or(Node::ACQUIRED, std::memory_order_release);
  }
```
would have the worker put the task for which it could not acquire all the semaphores back into the pool of tasks that can be reconsidered for execution later. This is what I would be expecting: that the worker that goes executing `f3_acquire_and_release_sema2` does not keep sitting on `f2_acquire_sema1_under_sema2` preventing its execution by the other worker when it is looking for work.

### Example 2

In example 2 the story is very similar. If you remove the module task box, what will happen initially goes along the same lines:

- the worker that started by executing `f1_acquire_sema1` will then likely execute `mod_f2` (no longer in a module, though) and will start looking for something else, as the direct successor (`mod_f3`) is not ready. And will steal `f5_acquire_sema1`. And as `sema1` is not available, it may even wait there, or go into a sleep state. This is not a problem now though, because neither `mod_f3` nor `f1_release_sema1` are blocked and will be ready for execution in due time;
- when the worker that finished doing `mod_f1` (or the one busy with `f4`, `f4` was a distraction to have things happen in the order I wanted) is done, it can continue with `mod_f3` and then with `f1_release_sema1`. When `sema1` is unblocked, `f5_acquire_sema1` becomes ready and its precondition is satisfiable, so some worker grabs it and executes it. 

So the issue is not due to the worker stealing some task which wants busy semaphores. The issue is because the particular thread keeps sitting on a task it did not execute and not letting other threads execute it later.

### Possible fix
I back off on my words now - it is not possible to make a blob dummy task happen after all the tasks of a subtree, if one does not know anything more about the structure of that subtree. In our case it was possible - just compile the AST of the DSL in use into taskflow subtrees with entry and exit tasks. But in general? There could be condition tasks and the like in the subtree, so making a dependency on all of them is no-go.

**Comment by ljwo** [comment]
So your changes to the module tasks of https://github.com/taskflow/taskflow/issues/649 solve issue 2 on the dev branch. Would you be able to do the same to semaphores to fix example 1?

**Comment by tsung-wei-huang** [comment]
@ljwo I am glad the second problem is solved. 

Regarding the first problem, semaphore is much more difficult to fix. Even the previous scheduling method won't solve this problem either as it relies on *luck* from the scheduler to escape from the deadlock. When a task is re-inserted into the queue due to an un-acquired semaphore, it is very likely the scheduler will immediately schedule that task to the same worker again. To leverage cache, Taskflow's worker queue operates in a stack fashion, with the last inserted tasks popped out first. 

As you pointed out with the following scheduling method,

```
   if(node->_semaphores && !node->_semaphores->to_acquire.empty()) {
    SmallVector<Node*> nodes;
    if(!node->_acquire_all(nodes)) {
      _schedule(worker, nodes);
      return;
    }
    node->_state.fetch_or(Node::ACQUIRED, std::memory_order_release);
  }
```

the statement "*would have the worker put the task for which it could not acquire all the semaphores back into the pool of tasks that can be reconsidered for execution later.*"  can still lead to a deadlock:  `f2_acquire_sema2`->`f2_acquire_sema1_under_sema2`->  can't acquire sema1 so steal-run `f3_acquire_and_release_sema2` -> can't acquire sema2, reinsert `f3_acquire_and_release_sema2` into queue, steal-run ``f3_acquire_and_release_sema2` again, etc. 

Also, the previous scheduling method is not context-aware since the context will be destroyed when the task is re-inserted into the queue. For example, the previous scheduling method can't handle the following scenario:

```
taskflow.emplace([](tf::Runtime& rt) {
   int local_state1 = do_something1();
   rt.acquire(semaphore1);
   int local_state2 = do_something2();
   rt.acquire(semaphore2);
   ....
   rt.release(semaphore1, semaphore2);
});
``` 

When the running worker fails to acquire `semaphore1`, we can't release it and ask the scheduler to re-consider this task later as this will cause the worker to lose the context (e.g., `local_state1`). 

Does this make sense to you?

**Comment by ljwo** [comment]
Sorry for the delay in my replying. I'll try to answer in points:

1. Regarding your later point, imho, this syntax:

```
taskflow.emplace([](tf::Runtime& rt) {
   int local_state1 = do_something1();
   rt.acquire(semaphore1);
   int local_state2 = do_something2();
   rt.acquire(semaphore2);
   ....
   rt.release(semaphore1, semaphore2);
});
```
is dead on arrival and cannot be fixed. This is user code, and as you say, nothing can be assumed about user code. And, again imho, this is the wrong thing to do. I'd expect a taskflow requiring two semaphores to be suitably structured in taskflow and not in user code. That is, this example would be rephrased as something like
```
    auto t1 = taskflow.emplace([](){int local_state1 = do_something1();});
    t1.acquire(semaphore1);

    auto t2 = taskflow.emplace([](){int local_state2 = do_something2();});
    t2.acquire(semaphore2);
    t1.precede(t2);

    auto t3 = taskflow.emplace( ... );;
    t2.precede(t3);

    t3.release(semaphore1, semaphore2);
```
Only this syntax lets `taskflow` actually **schedule** semaphored tasks, instead of just blatantly running them.

2. Regarding your earlier point, I don't quite agree. There is a point in time in which `sema1` gets released by the worker thread (call it A) that was busy with `f1_acquire_and_release_sema1` and that worker A is looking for work. So even if worker B that you mention keeps banging its head against `f3_acquire_and_release_sema2`, worker A will then happily (so I presume) execute `f2_acquire_sema1_under_sema2` and what follows. And when it is finally done with `f2_release_sema2`, the guy B that was repeatedly trying to steal `f3_acquire_and_release_sema2` suddenly can steal it and execute it. And no task waited longer than it actually needed to.

3. There is a third issue that I came to understand only recently. In the implementation of `v3.8.0` the semaphores suffer from the same performance hit with the same mechanism as the one we discussed at module tasks. So you enter a task, cannot acquire the semaphore because it is busy and go stealing. And happen to steal a `WaitForTimer@10minutes` that heats your samples in the machine. Now, your semaphore is released after some milliseconds, but you're out of luck. Frankly, this makes those semaphores pretty useless from my point of view. With module tasks gone (or fixed in the upcoming release) I'd rather use OS semaphores and have the worker thread just wait on the OS semaphore, so that it can be notified as soon as it can proceed, maybe bumping the number of worker threads in the executor by a few. But it makes me sad, the scheduler **should** be able to do a better job at that. I don't have a working example for that problem today, will be back next week with it.

**Comment by ljwo** [comment]
So the promised example is in the same repo at https://github.com/ljwo/taskflow-semaphores in namespace `time3`. The structure is as simple as that:
![dump-time3](https://github.com/user-attachments/assets/b952f277-dad2-4eeb-b9ff-feb8a4abc82d)
There is only one semaphore in this example, started at value 1. There are two workers that usually start executing `f0` and `f2_sema1`. The placeholder `f0` is introducing a slight delay (2ms), so that `f1_sema1` usually runs while `f2_sema1` has already started. Both tasks that require semaphores take roughly 10ms each to complete their work (once they acquire the semaphore). Since `sema1` cannot be acquired when `f1_sema1` is run, the worker goes stealing and grabs `f3`. There is still some time before `sema1` is released, so if `f3` runs fast (say, 5ms), we might not even notice any problem:
```
[worker_id:0] f0 starting work at 2701ns 
[worker_id:1] f2_sema1 acquiring sema1 at 2847ns 
[worker_id:1] f2_sema1 acquired sema1 at 57392ns starting work 
[worker_id:0] f0 work done, finished at 2123793ns 
[worker_id:0] f1_sema1 acquiring sema1 at 2145713ns 
[worker_id:0] f3 starting work at 2147446ns 
[worker_id:0] f3 work done, finished at 7221512ns 
[worker_id:1] f2_sema1 released sema1 finished at 10135629ns 
[worker_id:0] f1_sema1 acquired sema1 at 10135671ns starting work 
[worker_id:0] f1_sema1 released sema1 finished at 20232771ns 
finish
```
(The system clock reference point is taken before the taskflow is run, so the reported initial times have a small offset). 

So, worker 0 was able to finish executing `f3` in time before `f2_sema1` released `sema1`, and `f1_sema1` started roughly at the expected 10ms delay from the beginning. Good.

But what if `f3` is slow and takes, say, 500ms?
```
[worker_id:1] f0 starting work at 4559ns 
[worker_id:0] f2_sema1 acquiring sema1 at 5035ns 
[worker_id:0] f2_sema1 acquired sema1 at 19171ns starting work 
[worker_id:1] f0 work done, finished at 2066783ns 
[worker_id:1] f1_sema1 acquiring sema1 at 2070072ns 
[worker_id:1] f3 starting work at 2073295ns 
[worker_id:0] f2_sema1 released sema1 finished at 10083730ns 
[worker_id:1] f3 work done, finished at 502151538ns 
[worker_id:1] f1_sema1 acquired sema1 at 502208182ns starting work 
[worker_id:1] f1_sema1 released sema1 finished at 512284617ns 
finish
```
That's a catastrophe. Now, `f1_sema1` is blocked for the time it takes the completely unrelated `f3` to finish. The semaphore is released after `f2_sema1` finishes and remains available for 490ms before `f1_sema1` is back and able to acquire it and proceed, even though there is a free worker thread in the system that has nothing else to do. And the 500ms is just a convenient tease, the upper limit is what any synchronous api call in `f3` can take. Minutes, in my case.

This was the simplest example I came up with, but you may split the acquire/release operations and have something like the following:
![dump-time2](https://github.com/user-attachments/assets/089becca-683e-4220-9093-d6bc2e21193d)
implemented in namespace `time2` with the same behaviour. And I believe the old implementation of semaphores in taskflow, the one that required the user to specify `task.acquire(sema)` and `task.release(sema)` was free from this particular issue.

**Comment by tsung-wei-huang** [comment]
@ljwo I agreed with you that the runtime-based acquire/release is problematic due to corun. The best solution for it might be be C++ coroutine instead of blocking-based corun. I will disable its interface for now and bring back the previous `task.acquire` and `task.release` models. It will take a few days. At the same time, please feel free to organize your examples as unittests.

**Comment by tsung-wei-huang** [comment]
@ljwo could you please check the master branch? I have brought back the previous interface - which should resolve your problem. More details can be referred to [Limit the Maximum Concurrency](https://taskflow.github.io/taskflow/LimitTheMaximumConcurrency.html).

**Comment by tsung-wei-huang** [comment]
I assume this has been resolved as there are no more activities.
